source/opt/licm_pass.cpp

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <queue>
+#include <utility>
+
+#include "opt/licm_pass.h"
+#include "opt/module.h"
+#include "opt/pass.h"
+
+namespace spvtools {
+namespace opt {
+
+LICMPass::LICMPass() {}
+
+Pass::Status LICMPass::Process(ir::IRContext* context) {
+  bool modified = false;
+
+  if (context != nullptr) {
+    ir_context = context;
+    modified = ProcessIRContext();
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool LICMPass::ProcessIRContext() {
+  bool modified = false;
+  ir::Module* module = ir_context->module();
+
+  // Process each function in the module
+  for (ir::Function& f : *module) {
+    modified |= ProcessFunction(&f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessFunction(ir::Function* f) {
+  bool modified = false;
+  ir::LoopDescriptor loop_descriptor{f};
+
+  // Process each loop in the function
+  for (ir::Loop& loop : loop_descriptor) {
+    modified |= ProcessLoop(&loop, f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessLoop(ir::Loop* loop, ir::Function* f) {
+  // Process all nested loops first
+  for (ir::Loop* nested_loop : *loop) {
+    ProcessLoop(nested_loop, f);
+  }
+
+  std::queue<ir::Instruction*> loop_iv_instr{};
+  GatherAllImmediatelyInvariantInstructions(loop, &loop_iv_instr);
+
+  return ProcessInstructionList(loop, &loop_iv_instr);
+}
+
+void LICMPass::GatherAllImmediatelyInvariantInstructions(
+    ir::Loop* loop, std::queue<ir::Instruction*>* loop_iv_instr) {
+  // For each instruction in each basic block in the loop, check if it is
+  // invariant. If so, push it to our invariants queue.
+  for (uint32_t bb_id : loop->GetBlocks()) {
```

Comment:
If you already processed the inner loops, there is not need to go over the blocks in the nested loop again.  You can cut down on the amount of work you need to do here.